### PR TITLE
Fix parsing content of PKCS#7-message if it is DER-data

### PR DESCRIPTION
**Pull request by anthrax63**
If i use DER-data as a "content" of PKCS#7 message then after i parse it from PEM-format i get the ByteString with zero length instead of valid DER-bytes. This commit fixes that problem.

### CODE REVIEW

File: lib/pkcs7.js [hunk 1]
```diff
@@ -342,16 +342,20 @@ p7.createSignedData = function() {
         if('content' in msg) {
           var content;
           if(msg.content instanceof forge.util.ByteBuffer) {
-            content = msg.content.bytes();
+            content = asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OCTETSTRING, false,
```
Comment (dlongley) [inline lib/pkcs7.js]: I believe this would break the design here. The format for the `content` of PKCS#7 messages is dependent on the `contentType`. The current design allows for people to specify a `ByteBuffer` (or binary string of bytes) with DER-encoded content -- that expresses whatever ASN.1 format they desire. This would force the content to be an `asn1.Type.OCTETSTRING`, which is undesirable.

**Comment by dlongley** [comment]
I may have misread these changes ... looking back over it again.

File: lib/pkcs7.js [hunk 2]
```diff
@@ -342,16 +342,20 @@ p7.createSignedData = function() {
...
+              msg.content.bytes());
           } else if(typeof msg.content === 'string') {
-            content = forge.util.encodeUtf8(msg.content);
+            content = asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OCTETSTRING, false,
+              forge.util.encodeUtf8(msg.content));
+          } else if (typeof msg.content === 'object') {
+            content = asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OCTETSTRING, true,
+              [msg.content]);
```
Comment (dlongley) [inline lib/pkcs7.js]: So, looking back at this code, this section of code only executes as a helper to automatically generate `contentInfo` in a common case. I think there was a presumption that the `contentInfo` would be generated externally unless `msg.content` is set to a byte buffer or a string. However, the conditional seems to allow this code to execute if `msg.content` is an `object` ... and I'm not sure why yet. I'll need to see if there is well defined behavior in this case in the documentation, but it seems to me that `msg.content` as an `object` could imply a number of different things and we can't simply pick to encode it like this.

In other words, if your `msg.content` is a string or a byte buffer, we have a well defined way in which it will be automatically wrapper. If you need to do something else, you should generate your own `contentInfo`. If a simple tweaking to this code better handles some other common cases that's fine, but we need to say explicitly what's going to happen if you pass in `msg.content` as an object -- and it's not yet clear to me that what's going on here is the "right" thing (or the highest utility thing) to do.

Comment (dlongley) [inline lib/pkcs7.js]: I'm also not yet sure how my above comment meshes with parsing PKCS#7 messages, so I'll have to look back at that.

File: tests/unit/pkcs7.js [hunk 3]
```diff
@@ -518,6 +519,23 @@ var support = require('./support');
...
+      p7.content = UTIL.createBuffer(ASN1.toDer(ASN1.create(ASN1.Class.UNIVERSAL, ASN1.Type.OCTETSTRING, false, 'Just a little test')));
+      p7.encrypt();
+
+      var pem = PKCS7.messageToPem(p7);
+
+      // convert back from PEM to new PKCS#7 object, decrypt, and test
+      p7 = PKCS7.messageFromPem(pem);
+      p7.decrypt(p7.recipients[0], PKI.privateKeyFromPem(_pem.privateKey));
+      var asn1 = ASN1.fromDer(p7.content);
+      ASSERT.equal(asn1.value, 'Just a little test');
```
Comment (dlongley) [inline tests/unit/pkcs7.js]: This test case looks good from what I can tell. Will look more closely when I get a chance.

File: lib/pkcs7.js (hunk 1 above)
Comment (dlongley) [inline lib/pkcs7.js]: Nevermind -- I see that the OCTETSTRING below was, in effect, moved up here (this is not a new one that wraps it).

**Comment by GitStorageOne** [comment]
No activity since 2017?
